### PR TITLE
Add Bidirectional/Websocket connections

### PR DESCRIPTION
**Pull request by IonesioJunior**
## Description
- [x] Create BidirectionalConnection superclass
- [x] Move ServerConnection/ClientConnection to BidirectionalConnection
- [x] Create Async WebsocketConnection class extending BidirectionalConnection class
- [ ] Add tests

## Checklist
- [X] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [X] I have commented on my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [X] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

File: src/syft/core/io/connection.py [hunk 1]
```diff
@@ -8,7 +8,53 @@
...
+        raise NotImplementedError
+
+    @syft_decorator(typechecking=True)
+    def send_signed_msg_with_reply(self, msg: SignedMessage) -> SignedMessage:
+        raise NotImplementedError
+
+
 class ServerConnection(object):
+    def __init__(self):
+        self.opt_bidirectional_conn = BidirectionalConnection()
```
Comment (IonesioJunior) [inline src/syft/core/io/connection.py:56]: @iamtrask  like this?

Comment (iamtrask) [reply src/syft/core/io/connection.py:56]: perfect!

**Comment by madhavajay** [comment]
Hi @IonesioJunior awesome work!

I made a branch off this PR and fixed a few issues:
https://github.com/OpenMined/PySyft/tree/bidirectional_connection_fixes

You should be able to merge or rebase this into your branch and then push again.
I didn't get time for WS Duet tests, but that would be a good idea.
The WSDuet needs something to "serve" so it can "connect", thats as far as I got after all the little fixes and clean ups.

I also added some enhanced debugging by allowing naming to persist across serialization and within client and address creation. Also the ability to see whats in the store easily. I also added a random_name generator just like Docker, so we should always have something to see and match in Client / Node / Address both with the emoji Visual IDs / UUIDs and the Unique Names (if we don't set one).

Check out the Early Dev notebook:
Screen Shot 2020-08-18 at 3 09 43 pm

Also, I think we are trying to upgrade our `protoc` to a newer version so that the git commits don't always conflict.
The newer version creates this line in all the pb2 files:
Screen Shot 2020-08-18 at 3 12 17 pm

Can you check your version and try upgrading?
```
$ protoc --version
libprotoc 3.12.4
```

One last thing, it would be great if you could also use the pre_commit script during development, it helps catch flake8 and mypy issues early.

Just run it like this:
```
$ scripts/pre_commit.sh
```

I find its awesome to run before PR so that I know the linter and tests will all pass in CI.

**Comment by madhavajay** [comment]
Also @iamtrask  and @IonesioJunior, I found one possible issue I wanted to check:
`src/syft/core/node/common/service/child_node_lifecycle_service.py`
Screen Shot 2020-08-18 at 5 27 57 pm

Since we needed to change the storage to use serialized versions of the client addresses (the whole client wasn't available in the method to get its id and other properties), I added a field for the key / look up, and since in the case of Client it is usually the "address" you want to dispatch to, I had it set to:
`lookup_id=client.id`

But there could be other times where the storage key is something like a msg.id, so I hope I haven't broken anything, just double checking this is correct. Also speaking of, the "msg_id" is now persisted across the serialization barrier in and out of signing a message, which makes it super easy to track a message in the debug output, however... I don't know if there is an expectation that msg_id's are unique between node routing hops or between the non-signed and signed version of a message?

**Comment by IonesioJunior** [comment]
Woah!  @madhavajay I don't know what exactly you did, but it solved the major part of my issues. Thank you so much!

Also, I'd like to ask you a couple of questions that maybe you can help me to understand.
**First**: In your PR, you deleted the `obj2pointer_type` method, but in the current stage of `syft_0.3.0` branch we're still using it during `Search` service. Should I keep this method there, or should I delete it?

**Second**: I'm finally able to build new SignalingMessage schemas and serialize them, but during the deserialization process, I'm getting an error about the lack of `schema2type`  attribute. It happens when I try to deserialize one of my message attributes (string type) during proto2obj execution. Am I missing something?

Again, thank you so much for your help!  :)

**Comment by madhavajay** [comment]
> Woah! @madhavajay I don't know what exactly you did, but it solved the major part of my issues. Thank you so much!
> 
> Also, I'd like to ask you a couple of questions that maybe you can help me to understand.
> **First**: In your PR, you deleted the `obj2pointer_type` method, but in the current stage of `syft_0.3.0` branch we're still using it during `Search` service. Should I keep this method there, or should I delete it?
> 
You are 100% correct, this is a mistake and im glad you spotted it.

> **Second**: I'm finally able to build new SignalingMessage schemas and serialize them, but during the deserialization process, I'm getting an error about the lack of `schema2type` attribute. It happens when I try to deserialize one of my message attributes (string type) during proto2obj execution. Am I missing something?
> Again, thank you so much for your help! :)

I created a test and the fixed a few issues, and it works. I think the issue was you were not saving the data so when it tried to create a new object without the data on the deserialization side it failed.
Take a look at the changes and you will see what I did. Also its a good idea to make a serde test for any new messages or services and then you can be sure they are working.

I rebased your branch onto syft_0.3.0 so that it can be merged immediately. Unfortunately theres no easy way to PR to a PR, so for now we can just do it over here:
https://github.com/OpenMined/PySyft/pull/4046

Or if you prefer you can merge this into your branch, but because of the rebase the commits have been re-written.
Your commit history and "authorship" of the code is retained so there isnt really any downside to the rebase.

I will message you on slack a link to the new API Example notebook which might help explain a bit of the Serde / SignedMessage stuff. I think we should merge this ASAP and then continue to work on it branching back off syft_0.3.0 keeping the PR's to shorter 1 day cycles so that the require less rebasing. Since the Websocket stuff is fairly independent there shouldnt be much conflict anyway.

**Comment by IonesioJunior** [comment]
These features will be merged by #4046

**Pull request closed without merging**